File: weird_captcha_gym/shared_runtime/server/incubator_graders/lanternfin_dive.py
from __future__ import annotations

import copy
import math
from typing import Any


MECHANIC_ID = "lanternfin_dive"
CHANNELS = ("tail", "yaw", "pitch", "roll")


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _close(first: Any, second: Any, tolerance: float = 0.012) -> bool:
    try:
        return math.isfinite(float(first)) and math.isfinite(float(second)) and abs(float(first) - float(second)) <= tolerance
    except (TypeError, ValueError):
        return False


def _same_state(first: dict[str, Any], second: dict[str, Any], tolerance: float = 0.012) -> bool:
    return all(_close(first.get(key), second.get(key), tolerance) for key in second)


def _clamp(value: float, low: float = -1.0, high: float = 1.0) -> float:
    return max(low, min(high, value))


def _step(state: dict[str, float], controls: dict[str, float], physics: dict[str, Any]) -> None:
    dt = float(physics["dt"])
    tail = _clamp(float(controls["tail"]))
    yaw = _clamp(float(controls["yaw"]))
    pitch = _clamp(float(controls["pitch"]))
    roll = _clamp(float(controls["roll"]))
    state["tail_phase"] += dt * (3.2 + 0.8 * abs(tail))
    state["fin_phase"] += dt * (2.0 + 0.35 * (abs(pitch) + abs(roll)))
    angular_damping = float(physics["angular_damping"])
    torque_gain = float(physics["torque_gain"])
    state["yaw_rate"] = state["yaw_rate"] * angular_damping + yaw * torque_gain
    state["pitch_rate"] = state["pitch_rate"] * angular_damping + pitch * torque_gain
    state["roll_rate"] = state["roll_rate"] * angular_damping + roll * torque_gain
    state["yaw"] += state["yaw_rate"]
    state["pitch"] = max(-1.2, min(1.2, state["pitch"] + state["pitch_rate"]))
    state["roll"] = max(-1.2, min(1.2, state["roll"] + state["roll_rate"]))
    forward_x = math.cos(state["pitch"]) * math.cos(state["yaw"])
    forward_y = math.sin(state["pitch"])
    forward_z = math.cos(state["pitch"]) * math.sin(state["yaw"])
    stroke = tail * (0.56 + 0.44 * math.sin(state["tail_phase"]))
    thrust = float(physics["thrust_gain"]) * 0.1 * stroke
    # Fins make a small fluid-relative lift contribution as well as changing
    # attitude. This keeps all four visible channels coupled to the 3D motion.
    lift = float(physics["thrust_gain"]) * 0.018 * math.sin(state["fin_phase"])
    state["vx"] = state["vx"] * float(physics["linear_damping"]) + forward_x * thrust
    state["vy"] = state["vy"] * float(physics["linear_damping"]) + forward_y * thrust + pitch * lift
    state["vz"] = state["vz"] * float(physics["linear_damping"]) + forward_z * thrust + roll * lift
    state["x"] += state["vx"]
    state["y"] += state["vy"]
    state["z"] += state["vz"]


def _arrival(state: dict[str, float], target: dict[str, Any], physics: dict[str, Any]) -> bool:
    distance = math.sqrt(sum((state[axis] - float(target[axis])) ** 2 for axis in ("x", "y", "z")))
    speed = math.sqrt(sum(state[axis] ** 2 for axis in ("vx", "vy", "vz")))
    return (
        distance <= float(physics["arrival_radius"])
        and speed <= float(physics["speed_limit"])
        and abs(state["pitch"]) <= float(physics["upright_tolerance"])
        and abs(state["roll"]) <= float(physics["upright_tolerance"])
    )


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    if payload.get("mechanic_id") != MECHANIC_ID or truth.get("mechanic_id") != MECHANIC_ID:
        return _fail("mechanic mismatch")
    if payload.get("task_id") != truth.get("task_id") or payload.get("challenge_id") != truth.get("challenge_id") or public.get("challenge_id") != truth.get("challenge_id"):
        return _fail("stale task or challenge")
    for key in ("initial", "target", "physics"):
        if public.get(key) != truth.get(key):
            return _fail(f"public {key} differs from the challenge contract")
    condition = truth.get("control_condition")
    if condition is not None and public.get("control_condition") != condition:
        return _fail("Lanternfin control condition mismatch")
    interaction = str((condition or {}).get("interaction") or "")
    expected_sources = None if condition is None else {
        "simplified": {"torque_button", "torque_neutral_button"},
        "full": {"torque_keyboard", "torque_neutral_keyboard"},
    }.get(interaction)
    if condition is not None and expected_sources is None:
        return _fail("Lanternfin interaction mode is invalid")
    events = payload.get("events")
    if not isinstance(events, list) or len(events) > 7000:
        return _fail("Lanternfin transcript malformed")
    state = {key: float(value) for key, value in truth["initial"].items()}
    controls = {channel: 0.0 for channel in CHANNELS}
    physics = truth["physics"]
    target = truth["target"]
    tick = 0
    hold = 0
    terminal = None
    for sequence, item in enumerate(events, 1):
        if not isinstance(item, dict) or item.get("seq") != sequence:
            return _fail(f"event {sequence} sequence invalid")
        kind = item.get("type")
        if terminal is not None:
            return _fail("events appeared after pearl certification")
        if kind == "abandon":
            return _fail("dive abandoned")
        if kind == "camera":
            continue
        if kind not in {"torque", "certify"}:
            return _fail(f"unknown Lanternfin event {kind!r}")
        try:
            event_tick = int(item.get("tick"))
        except (TypeError, ValueError):
            return _fail("Lanternfin event has no physical tick")
        if event_tick < tick or event_tick > int(physics["max_ticks"]):
            return _fail("Lanternfin event time moved backward or beyond the dive")
        for _ in range(tick, event_tick):
            _step(state, controls, physics)
            hold = hold + 1 if _arrival(state, target, physics) else 0
        tick = event_tick
        if kind == "torque":
            channel = str(item.get("channel") or "")
            if channel not in CHANNELS:
                return _fail("unknown torque channel")
            if expected_sources is not None and item.get("input_source") not in expected_sources:
                return _fail("torque used the wrong interaction input")
            if not _same_state(item.get("before") or {}, {channel: controls[channel]}, 0.02):
                return _fail("torque event started from a stale control value")
            try:
                value = float(item.get("value"))
            except (TypeError, ValueError):
                return _fail("torque value is not numeric")
            if value not in {-1.0, 0.0, 1.0}:
                return _fail("torque value is outside the visible detents")
            controls[channel] = value
            try:
                after = float(item.get("after"))
            except (TypeError, ValueError):
                return _fail("torque event has no after detent")
            if after != value:
                return _fail("torque event reports a false detent")
        else:
            terminal = item
            if not _same_state(terminal.get("state") or {}, state, 0.018):
                return _fail("certification state disagrees with fluid replay")
            if int(terminal.get("hold_ticks", -1)) != hold:
                return _fail("upright arrival hold disagrees with replay")
            accepted = hold >= int(physics["hold_ticks"]) and _arrival(state, target, physics)
            if bool(terminal.get("accepted")) != accepted:
                return _fail("visible arrival verdict disagrees with replay")
    if not isinstance(terminal, dict):
        return _fail("no pearl certification")
    accepted = bool(terminal.get("accepted"))
    passed = accepted and payload.get("completed") is True
    distance = math.sqrt(sum((state[axis] - float(target[axis])) ** 2 for axis in ("x", "y", "z")))
    speed = math.sqrt(sum(state[axis] ** 2 for axis in ("vx", "vy", "vz")))
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"3D fluid replay reached pearl: distance={distance:.3f}, speed={speed:.3f}, upright hold={hold}"
            if passed
            else f"pearl not held: distance={distance:.3f}, speed={speed:.3f}, upright hold={hold}/{physics['hold_ticks']}"
        ),
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/lanternfin_dive.js
(() => {
  "use strict";
  const esc = (value) => String(value ?? "").replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const copy = (value) => JSON.parse(JSON.stringify(value));
  const clamp = (value, low = -1, high = 1) => Math.max(low, Math.min(high, value));
  const round = (value, places = 3) => Math.round(Number(value) * 10 ** places) / 10 ** places;
  const channels = ["tail", "yaw", "pitch", "roll"];
  const labels = {tail: "TAIL BEAT", yaw: "YAW FIN", pitch: "PITCH FIN", roll: "ROLL FIN"};
  const keyMap = {
    KeyA: ["tail", -1], KeyD: ["tail", 1],
    KeyJ: ["yaw", -1], KeyL: ["yaw", 1],
    KeyI: ["pitch", 1], KeyK: ["pitch", -1],
    KeyQ: ["roll", -1], KeyE: ["roll", 1],
  };

  function step(fish, controls, physics) {
    const dt = Number(physics.dt);
    const tail = clamp(Number(controls.tail));
    const yaw = clamp(Number(controls.yaw));
    const pitch = clamp(Number(controls.pitch));
    const roll = clamp(Number(controls.roll));
    fish.tail_phase += dt * (3.2 + 0.8 * Math.abs(tail));
    fish.fin_phase += dt * (2.0 + 0.35 * (Math.abs(pitch) + Math.abs(roll)));
    fish.yaw_rate = fish.yaw_rate * Number(physics.angular_damping) + yaw * Number(physics.torque_gain);
    fish.pitch_rate = fish.pitch_rate * Number(physics.angular_damping) + pitch * Number(physics.torque_gain);
    fish.roll_rate = fish.roll_rate * Number(physics.angular_damping) + roll * Number(physics.torque_gain);
    fish.yaw += fish.yaw_rate;
    fish.pitch = clamp(fish.pitch + fish.pitch_rate, -1.2, 1.2);
    fish.roll = clamp(fish.roll + fish.roll_rate, -1.2, 1.2);
    const forward = {
      x: Math.cos(fish.pitch) * Math.cos(fish.yaw),
      y: Math.sin(fish.pitch),
      z: Math.cos(fish.pitch) * Math.sin(fish.yaw),
    };
    const stroke = tail * (0.56 + 0.44 * Math.sin(fish.tail_phase));
    const thrust = Number(physics.thrust_gain) * 0.1 * stroke;
    const lift = Number(physics.thrust_gain) * 0.018 * Math.sin(fish.fin_phase);
    fish.vx = fish.vx * Number(physics.linear_damping) + forward.x * thrust;
    fish.vy = fish.vy * Number(physics.linear_damping) + forward.y * thrust + pitch * lift;
    fish.vz = fish.vz * Number(physics.linear_damping) + forward.z * thrust + roll * lift;
    fish.x += fish.vx;
    fish.y += fish.vy;
    fish.z += fish.vz;
  }

  function distance(fish, target) {
    return Math.hypot(fish.x - target.x, fish.y - target.y, fish.z - target.z);
  }

  function arrival(fish, target, physics) {
    return distance(fish, target) <= Number(physics.arrival_radius)
      && Math.hypot(fish.vx, fish.vy, fish.vz) <= Number(physics.speed_limit)
      && Math.abs(fish.pitch) <= Number(physics.upright_tolerance)
      && Math.abs(fish.roll) <= Number(physics.upright_tolerance);
  }

  function project(point, camera, width, height) {
    let x = Number(point.x) - camera.focus.x;
    let y = Number(point.y) - camera.focus.y;
    let z = Number(point.z) - camera.focus.z;
    const cy = Math.cos(camera.yaw), sy = Math.sin(camera.yaw);
    const rotatedX = cy * x + sy * z;
    const rotatedZ = -sy * x + cy * z;
    const cp = Math.cos(camera.pitch), sp = Math.sin(camera.pitch);
    const vertical = cp * y - sp * rotatedZ;
    const depth = sp * y + cp * rotatedZ + 3.8;
    const scale = 230 / Math.max(1.0, depth);
    return {x: width * 0.5 + rotatedX * scale, y: height * 0.56 - vertical * scale, depth, scale};
  }

  function fishBasis(fish) {
    const forward = {x: Math.cos(fish.pitch) * Math.cos(fish.yaw), y: Math.sin(fish.pitch), z: Math.cos(fish.pitch) * Math.sin(fish.yaw)};
    const right = {x: -Math.sin(fish.yaw), y: 0, z: Math.cos(fish.yaw)};
    const up = {x: -Math.sin(fish.pitch) * Math.cos(fish.yaw), y: Math.cos(fish.pitch), z: -Math.sin(fish.pitch) * Math.sin(fish.yaw)};
    return {forward, right, up};
  }

  function renderFish(ctx, fish, camera, width, height) {
    const basis = fishBasis(fish);
    const add = (scale, direction = basis.forward) => ({x: fish.x + direction.x * scale, y: fish.y + direction.y * scale, z: fish.z + direction.z * scale});
    const nose = project(add(0.23), camera, width, height);
    const tail = project(add(-0.24), camera, width, height);
    const body = project({x: fish.x, y: fish.y, z: fish.z}, camera, width, height);
    const finLeft = project({x: fish.x + basis.right.x * 0.22, y: fish.y + basis.right.y * 0.22, z: fish.z + basis.right.z * 0.22}, camera, width, height);
    const finRight = project({x: fish.x - basis.right.x * 0.22, y: fish.y - basis.right.y * 0.22, z: fish.z - basis.right.z * 0.22}, camera, width, height);
    ctx.save();
    ctx.shadowBlur = 25;
    ctx.shadowColor = "#7cf7e8";
    const gradient = ctx.createLinearGradient(tail.x, tail.y, nose.x, nose.y);
    gradient.addColorStop(0, "#4c78cf");
    gradient.addColorStop(0.52, "#42ddd0");
    gradient.addColorStop(1, "#fff1a8");
    ctx.strokeStyle = gradient;
    ctx.lineWidth = Math.max(18, body.scale * 0.085);
    ctx.lineCap = "round";
    ctx.beginPath(); ctx.moveTo(tail.x, tail.y); ctx.lineTo(nose.x, nose.y); ctx.stroke();
    ctx.shadowBlur = 10;
    ctx.strokeStyle = "#bbfff2";
    ctx.lineWidth = Math.max(6, body.scale * 0.035);
    ctx.beginPath(); ctx.moveTo(finLeft.x, finLeft.y); ctx.lineTo(body.x, body.y); ctx.lineTo(finRight.x, finRight.y); ctx.stroke();
    ctx.strokeStyle = "#d6e7ff";
    ctx.lineWidth = Math.max(4, body.scale * 0.022);
    const tailTip = project(add(-0.42), camera, width, height);
    ctx.beginPath(); ctx.moveTo(tail.x, tail.y); ctx.lineTo(tailTip.x, tailTip.y); ctx.stroke();
    ctx.fillStyle = "#fff6c5"; ctx.beginPath(); ctx.arc(nose.x, nose.y, Math.max(4, body.scale * 0.027), 0, Math.PI * 2); ctx.fill();
    ctx.restore();
  }

  function shell(state, body, controls) {
    const condition = state.control_condition || {};
    const interaction = condition.interaction || "full";
    return `<section class="lanternfin" data-interaction="${esc(interaction)}" data-challenge-id="${esc(state.challenge_id)}">
      <header class="lanternfin-head"><div><span>ABYSSAL NAVIGATION AUTHORITY</span><h1>LANTERNFIN DIVE</h1></div><p>${esc(state.prompt)}</p></header>
      <main class="lanternfin-main">${body}<aside class="lanternfin-console">${controls}</aside></main>
      <footer class="lanternfin-foot"><span>${esc(state.challenge_id.toUpperCase())} · ${esc(interaction.toUpperCase())} INPUT SURFACE</span><div id="lanternfin-readout" class="readout" data-status="idle">OBSERVE THE WATER</div></footer>
    </section>`;
  }

  function controlMarkup(interaction) {
    if (interaction === "simplified") {
      return `<h2>JOINT TORQUES</h2><p class="lf-help">Each detent persists until you reverse or neutralize it. Watch the next frame before choosing the next stroke.</p><div class="lf-buttons">${channels.map((channel) => `<div class="lf-row"><b>${labels[channel]}</b><button data-channel="${channel}" data-value="-1">−</button><button data-channel="${channel}" data-value="0">NEUTRAL</button><button data-channel="${channel}" data-value="1">+</button></div>`).join("")}</div><p class="lf-help">The canvas is also an orbit camera: drag it to inspect the pearl's depth.</p>`;
    }
    return `<h2>HELD TORQUE KEYS</h2><p class="lf-help">Hold a key pair and release to neutral. A, D tail · J, L yaw · K, I pitch · Q, E roll.</p><div class="lf-keybank">${[["A / D", "TAIL BEAT"], ["J / L", "YAW FIN"], ["K / I", "PITCH FIN"], ["Q / E", "ROLL FIN"]].map((item) => `<div><kbd>${item[0]}</kbd><span>${item[1]}</span></div>`).join("")}</div><p class="lf-help">Drag the aquarium view to inspect the fish against the far wall and pearl.</p>`;
  }

  function render(state, helpers) {
    if (window.lanternfinDiveModel?.cleanup) window.lanternfinDiveModel.cleanup();
    document.body.dataset.mechanic = "lanternfin-dive";
    const condition = state.control_condition || {};
    const interaction = condition.interaction || "full";
    const model = {
      state, helpers, interaction, fish: copy(state.initial), controls: {tail: 0, yaw: 0, pitch: 0, roll: 0},
      events: [], tick: 0, hold: 0, terminal: false, submitting: false, timer: null,
      camera: {yaw: -0.48, pitch: 0.22, focus: {x: 0, y: 0, z: 0}}, drag: null, keys: new Map(),
    };
    window.lanternfinDiveModel = model;
    const buttons = controlMarkup(interaction);
    helpers.app.innerHTML = shell(state, `<div class="lanternfin-tank"><canvas id="lanternfin-canvas" width="760" height="520" aria-label="3D aquarium showing the lanternfin fish and destination pearl"></canvas><div class="lanternfin-legend"><i></i>PEARL · <b>amber</b> = depth cue · <b>cyan</b> = body orientation</div></div>`, buttons + `<div class="lf-telemetry"><div><span>POSITION</span><b id="lf-pos">0.00 / 0.00 / 0.00</b></div><div><span>ATTITUDE</span><b id="lf-att">ROLL 0° · PITCH 0°</b></div><div><span>PEARL RANGE</span><b id="lf-range">—</b></div><div><span>UPRIGHT HOLD</span><b id="lf-hold">0 / ${Number(state.physics.hold_ticks)}</b></div></div><button id="lf-certify" class="lf-primary">CERTIFY PEARL ARRIVAL</button><button id="lf-abandon" class="lf-danger">ABANDON / FRESH DIVE</button>`);
    const canvas = document.getElementById("lanternfin-canvas");
    const readout = document.getElementById("lanternfin-readout");
    const setMessage = (message, status = "idle") => { if (model.terminal && message !== "PASS") return; readout.textContent = message; readout.dataset.status = status; };
    const event = (type, details = {}) => { model.events.push({seq: model.events.length + 1, type, ...details}); };
    const stateSnapshot = () => copy(model.fish);
    const draw = () => {
      const ctx = canvas.getContext("2d");
      const width = canvas.width, height = canvas.height;
      ctx.clearRect(0, 0, width, height);
      const background = ctx.createLinearGradient(0, 0, width, height);
      background.addColorStop(0, state.palette === "amberglass" ? "#081c2c" : "#061d2b");
      background.addColorStop(1, "#122b3c"); ctx.fillStyle = background; ctx.fillRect(0, 0, width, height);
      ctx.strokeStyle = "rgba(96,205,217,.19)"; ctx.lineWidth = 1;
      for (let i = -2; i <= 2; i += 0.5) { const a = project({x: i, y: -0.78, z: -1.6}, model.camera, width, height); const b = project({x: i, y: -0.78, z: 1.8}, model.camera, width, height); ctx.beginPath(); ctx.moveTo(a.x, a.y); ctx.lineTo(b.x, b.y); ctx.stroke(); }
      for (let z = -1.6; z <= 1.8; z += 0.5) { const a = project({x: -2.0, y: -0.78, z}, model.camera, width, height); const b = project({x: 2.0, y: -0.78, z}, model.camera, width, height); ctx.beginPath(); ctx.moveTo(a.x, a.y); ctx.lineTo(b.x, b.y); ctx.stroke(); }
      const targetPoint = project(state.target, model.camera, width, height);
      ctx.save(); ctx.shadowBlur = 24; ctx.shadowColor = "#ffcc66"; ctx.strokeStyle = "#ffd477"; ctx.lineWidth = 4; ctx.beginPath(); ctx.arc(targetPoint.x, targetPoint.y, Math.max(11, targetPoint.scale * state.target.radius), 0, Math.PI * 2); ctx.stroke(); ctx.fillStyle = "rgba(255,190,82,.2)"; ctx.fill(); ctx.restore();
      const targetLabel = project({x: state.target.x, y: state.target.y + 0.16, z: state.target.z}, model.camera, width, height); ctx.fillStyle = "#ffe7a3"; ctx.font = "700 12px ui-monospace, monospace"; ctx.fillText("DESTINATION PEARL", targetLabel.x - 54, targetLabel.y);
      renderFish(ctx, model.fish, model.camera, width, height);
      ctx.fillStyle = "rgba(198,246,243,.75)"; ctx.font = "700 12px ui-monospace, monospace"; ctx.fillText(`TANK TIME ${model.tick} · VIEW ${Math.round(model.camera.yaw * 57.3)}°`, 18, 24);
      document.getElementById("lf-pos").textContent = `${model.fish.x.toFixed(2)} / ${model.fish.y.toFixed(2)} / ${model.fish.z.toFixed(2)}`;
      document.getElementById("lf-att").textContent = `ROLL ${(model.fish.roll * 57.3).toFixed(0)}° · PITCH ${(model.fish.pitch * 57.3).toFixed(0)}°`;
      document.getElementById("lf-range").textContent = `${distance(model.fish, state.target).toFixed(3)} m`;
      document.getElementById("lf-hold").textContent = `${model.hold} / ${Number(state.physics.hold_ticks)}`;
      channels.forEach((channel) => document.querySelector(`[data-value-for="${channel}"]`)?.replaceChildren(document.createTextNode(model.controls[channel] === 0 ? "NEUTRAL" : model.controls[channel] > 0 ? "+1" : "−1")));
    };
    const setTorque = (channel, value, inputSource) => {
      if (model.terminal || model.submitting || !channels.includes(channel)) return;
      value = Number(value);
      if (![-1, 0, 1].includes(value)) return;
      if (model.controls[channel] === value) { draw(); return; }
      event("torque", {tick: model.tick, channel, value, before: {[channel]: model.controls[channel]}, after: value, input_source: inputSource});
      model.controls[channel] = value;
      setMessage(`${labels[channel]} ${value === 0 ? "NEUTRAL" : value > 0 ? "+1" : "−1"}`, "active"); draw();
    };
    const tick = () => {
      if (model.terminal || model.submitting || model.tick >= Number(state.physics.max_ticks)) return;
      step(model.fish, model.controls, state.physics); model.tick += 1;
      model.hold = arrival(model.fish, state.target, state.physics) ? model.hold + 1 : 0;
      if (model.hold > 0) setMessage(`PEARL WINDOW · UPRIGHT ${model.hold}/${Number(state.physics.hold_ticks)}`, "passed");
      else if (model.tick >= Number(state.physics.max_ticks)) setMessage("DIVE WINDOW CLOSED · CERTIFY OR ABANDON", "error");
      draw();
    };
    const submit = async (payload) => {
      if (model.submitting || model.terminal) return;
      model.submitting = true; setMessage("INDEPENDENT 3D REPLAY IN PROGRESS…", "pending");
      try {
        const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
        const result = await response.json();
        if (result.passed === true) {
          model.terminal = true; setMessage("PASS", "passed");
          helpers.app.querySelector(".lanternfin")?.insertAdjacentHTML("beforeend", `<div class="lf-verdict is-pass"><strong>PEARL ARRIVAL ACCEPTED</strong><span>3D FLUID REPLAY · UPRIGHT HOLD · ORDINARY INPUT TRANSCRIPT</span></div>`);
        } else if (result.passed === false && result.state) {
          await helpers.render(result.state);
          helpers.app.querySelector(".lanternfin")?.insertAdjacentHTML("beforeend", `<div class="lf-verdict is-fresh"><strong>FAIL · FRESH DIVE</strong><span>${esc(result.feedback || "The failed dive was replaced")}</span></div>`);
        } else { model.submitting = false; setMessage(`FAIL · ${result.feedback || result.error || "NO AUTHORITATIVE GRADE"}`, "error"); }
      } catch (error) { model.submitting = false; setMessage(`FAIL · VERIFIER OFFLINE · ${error.message}`, "error"); }
    };
    if (interaction === "simplified") {
      document.querySelectorAll("[data-channel]").forEach((button) => button.addEventListener("click", () => setTorque(button.dataset.channel, Number(button.dataset.value), Number(button.dataset.value) === 0 ? "torque_neutral_button" : "torque_button")));
    } else {
      model.keyHandler = (keyboardEvent) => {
        const mapped = keyMap[keyboardEvent.code]; if (!mapped || keyboardEvent.repeat || model.terminal || model.submitting) return;
        keyboardEvent.preventDefault(); model.keys.set(keyboardEvent.code, true); setTorque(mapped[0], mapped[1], "torque_keyboard");
      };
      model.keyUpHandler = (keyboardEvent) => {
        const mapped = keyMap[keyboardEvent.code]; if (!mapped || !model.keys.has(keyboardEvent.code)) return;
        keyboardEvent.preventDefault(); model.keys.delete(keyboardEvent.code); setTorque(mapped[0], 0, "torque_neutral_keyboard");
      };
      document.addEventListener("keydown", model.keyHandler); document.addEventListener("keyup", model.keyUpHandler);
    }
    canvas.addEventListener("pointerdown", (pointerEvent) => { if (model.terminal || model.submitting) return; model.drag = {id: pointerEvent.pointerId, x: pointerEvent.clientX, y: pointerEvent.clientY}; canvas.setPointerCapture(pointerEvent.pointerId); });
    canvas.addEventListener("pointermove", (pointerEvent) => { if (!model.drag || model.drag.id !== pointerEvent.pointerId) return; const dx = pointerEvent.clientX - model.drag.x, dy = pointerEvent.clientY - model.drag.y; model.drag.x = pointerEvent.clientX; model.drag.y = pointerEvent.clientY; model.camera.yaw += dx * 0.008; model.camera.pitch = clamp(model.camera.pitch + dy * 0.006, -0.55, 0.65); draw(); });
    canvas.addEventListener("pointerup", (pointerEvent) => { if (!model.drag || model.drag.id !== pointerEvent.pointerId) return; model.drag = null; if (canvas.hasPointerCapture(pointerEvent.pointerId)) canvas.releasePointerCapture(pointerEvent.pointerId); });
    document.getElementById("lf-certify").addEventListener("click", () => {
      const accepted = model.hold >= Number(state.physics.hold_ticks) && arrival(model.fish, state.target, state.physics);
      event("certify", {tick: model.tick, state: stateSnapshot(), hold_ticks: model.hold, accepted, input_source: "certify_button"});
      if (!accepted) { void submit({mechanic_id: state.mechanic_id, task_id: state.task_id, challenge_id: state.challenge_id, events: model.events, completed: false}); return; }
      void submit({mechanic_id: state.mechanic_id, task_id: state.task_id, challenge_id: state.challenge_id, events: model.events, completed: true});
    });
    document.getElementById("lf-abandon").addEventListener("click", () => void submit({mechanic_id: state.mechanic_id, task_id: state.task_id, challenge_id: state.challenge_id, events: [...model.events, {seq: model.events.length + 1, type: "abandon"}], completed: false}));
    model.timer = setInterval(tick, Number(state.physics.tick_ms));
    model.cleanup = () => { if (model.timer) clearInterval(model.timer); if (model.keyHandler) document.removeEventListener("keydown", model.keyHandler); if (model.keyUpHandler) document.removeEventListener("keyup", model.keyUpHandler); model.terminal = true; };
    draw();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.lanternfin_dive = {rootSelector: ".lanternfin", render};
})();
